// viem/_cjs/chains/definitions/sepolia.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sepolia = void 0;
const defineChain_js_1 = require("../../utils/chain/defineChain.js");
exports.sepolia = (0, defineChain_js_1.defineChain)({
    id: 11155111,
    name: 'Sepolia',
    nativeCurrency: { name: 'Sepolia Ether', symbol: 'SEP', decimals: 18 },
    rpcUrls: {
        default: {
            http: ['https://rpc.sepolia.org'],
        },
    },
    blockExplorers: {
        default: {
            name: 'Etherscan',
            url: 'https://sepolia.etherscan.io',
            apiUrl: 'https://api-sepolia.etherscan.io/api',
        },
    },
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 751532,
        },
        ensRegistry: { address: '0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e' },
        ensUniversalResolver: {
            address: '0xBaBC7678D7A63104f1658c11D6AE9A21cdA09725',
            blockCreated: 5043334,
        },
    },
    testnet: true,
});
