// viem/_cjs/chains/opStack/actions/getWithdrawalStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWithdrawalStatus = void 0;
const readContract_js_1 = require("../../../actions/public/readContract.js");
const contract_js_1 = require("../../../errors/contract.js");
const abis_js_1 = require("../abis.js");
const withdrawal_js_1 = require("../errors/withdrawal.js");
const getWithdrawals_js_1 = require("../utils/getWithdrawals.js");
const getL2Output_js_1 = require("./getL2Output.js");
const getTimeToFinalize_js_1 = require("./getTimeToFinalize.js");
async function getWithdrawalStatus(client, parameters) {
    const { chain = client.chain, receipt, targetChain } = parameters;
    const portalAddress = (() => {
        if (parameters.portalAddress)
            return parameters.portalAddress;
        if (chain)
            return targetChain.contracts.portal[chain.id].address;
        return Object.values(targetChain.contracts.portal)[0].address;
    })();
    const [withdrawal] = (0, getWithdrawals_js_1.getWithdrawals)(receipt);
    if (!withdrawal)
        throw new withdrawal_js_1.ReceiptContainsNoWithdrawalsError({
            hash: receipt.transactionHash,
        });
    const [outputResult, proveResult, finalizedResult, timeToFinalizeResult] = await Promise.allSettled([
        (0, getL2Output_js_1.getL2Output)(client, {
            ...parameters,
            l2BlockNumber: receipt.blockNumber,
        }),
        (0, readContract_js_1.readContract)(client, {
            abi: abis_js_1.portalAbi,
            address: portalAddress,
            functionName: 'provenWithdrawals',
            args: [withdrawal.withdrawalHash],
        }),
        (0, readContract_js_1.readContract)(client, {
            abi: abis_js_1.portalAbi,
            address: portalAddress,
            functionName: 'finalizedWithdrawals',
            args: [withdrawal.withdrawalHash],
        }),
        (0, getTimeToFinalize_js_1.getTimeToFinalize)(client, {
            ...parameters,
            withdrawalHash: withdrawal.withdrawalHash,
        }),
    ]);
    if (outputResult.status === 'rejected') {
        const error = outputResult.reason;
        if (error.cause instanceof contract_js_1.ContractFunctionRevertedError &&
            error.cause.data?.args?.[0] ===
                'L2OutputOracle: cannot get output for a block that has not been proposed')
            return 'waiting-to-prove';
        throw error;
    }
    if (proveResult.status === 'rejected')
        throw proveResult.reason;
    if (finalizedResult.status === 'rejected')
        throw finalizedResult.reason;
    if (timeToFinalizeResult.status === 'rejected')
        throw timeToFinalizeResult.reason;
    const [_, proveTimestamp] = proveResult.value;
    if (!proveTimestamp)
        return 'ready-to-prove';
    const finalized = finalizedResult.value;
    if (finalized)
        return 'finalized';
    const { seconds } = timeToFinalizeResult.value;
    return seconds > 0 ? 'waiting-to-finalize' : 'ready-to-finalize';
}
exports.getWithdrawalStatus = getWithdrawalStatus;
