// viem/_cjs/chains/opStack/decorators/publicL1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.publicActionsL1 = void 0;
const buildInitiateWithdrawal_js_1 = require("../actions/buildInitiateWithdrawal.js");
const estimateDepositTransactionGas_js_1 = require("../actions/estimateDepositTransactionGas.js");
const estimateFinalizeWithdrawalGas_js_1 = require("../actions/estimateFinalizeWithdrawalGas.js");
const estimateProveWithdrawalGas_js_1 = require("../actions/estimateProveWithdrawalGas.js");
const getL2Output_js_1 = require("../actions/getL2Output.js");
const getTimeToFinalize_js_1 = require("../actions/getTimeToFinalize.js");
const getTimeToNextL2Output_js_1 = require("../actions/getTimeToNextL2Output.js");
const getTimeToProve_js_1 = require("../actions/getTimeToProve.js");
const getWithdrawalStatus_js_1 = require("../actions/getWithdrawalStatus.js");
const waitForNextL2Output_js_1 = require("../actions/waitForNextL2Output.js");
const waitToFinalize_js_1 = require("../actions/waitToFinalize.js");
const waitToProve_js_1 = require("../actions/waitToProve.js");
function publicActionsL1() {
    return (client) => {
        return {
            buildInitiateWithdrawal: (args) => (0, buildInitiateWithdrawal_js_1.buildInitiateWithdrawal)(client, args),
            estimateDepositTransactionGas: (args) => (0, estimateDepositTransactionGas_js_1.estimateDepositTransactionGas)(client, args),
            estimateFinalizeWithdrawalGas: (args) => (0, estimateFinalizeWithdrawalGas_js_1.estimateFinalizeWithdrawalGas)(client, args),
            estimateProveWithdrawalGas: (args) => (0, estimateProveWithdrawalGas_js_1.estimateProveWithdrawalGas)(client, args),
            getL2Output: (args) => (0, getL2Output_js_1.getL2Output)(client, args),
            getTimeToFinalize: (args) => (0, getTimeToFinalize_js_1.getTimeToFinalize)(client, args),
            getTimeToNextL2Output: (args) => (0, getTimeToNextL2Output_js_1.getTimeToNextL2Output)(client, args),
            getTimeToProve: (args) => (0, getTimeToProve_js_1.getTimeToProve)(client, args),
            getWithdrawalStatus: (args) => (0, getWithdrawalStatus_js_1.getWithdrawalStatus)(client, args),
            waitForNextL2Output: (args) => (0, waitForNextL2Output_js_1.waitForNextL2Output)(client, args),
            waitToFinalize: (args) => (0, waitToFinalize_js_1.waitToFinalize)(client, args),
            waitToProve: (args) => (0, waitToProve_js_1.waitToProve)(client, args),
        };
    };
}
exports.publicActionsL1 = publicActionsL1;
